google-oauth-client/src/main/java/com/google/api/client/auth/oauth/OAuthHmacSha256Signer.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright (c) 2010 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.api.client.auth.oauth;
+
+import com.google.api.client.util.Base64;
+import com.google.api.client.util.Beta;
+import com.google.api.client.util.StringUtils;
+import java.security.GeneralSecurityException;
+import javax.crypto.Mac;
+import javax.crypto.SecretKey;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * {@link Beta} <br>
+ * OAuth {@code "HMAC-SHA256"} signature method.
+ */
+@Beta
+public final class OAuthHmacSha256Signer implements OAuthSigner {
+
+  /** Client-shared secret or {@code null} for none. */
+  public String clientSharedSecret;
```

Comment:
private on fields please

Reply:
I can make these fields private, but then will also need to add accessors/mutators.  I noticed in the original class (OAuthHmacSigner) these fields were public and had methods to get/set the values.  Do we want to have this signing class implemented differently from the other on that it is based on?  Existing code using OAuthHmacSigner that needs to migrate to this class would have fewer code changes keeping it the same, but then the API was labelled Beta.

Reply:
mutable public fields are a hard no. Don't try to be consistent with bad code. You can probably make these final with no mutators.

Reply:
The client secret and the token secret probably shouldn't be final, as there is often a multi-step flow to obtain the values in separate steps, so they need to be set after the signer instance is created in most instances.

Also, in looking at the other code required for making a call, the `OAuthParameters` class will still have mutable public fields (similar to `OAuthHmacSigner`), and it is necessary to use that class together with the signer, to make a complete OAuth 1.0 signed request:

Current code needed for HMAC-SHA1:
```
OAuthHmacSigner hmacSha1Signer = new OAuthHmacSigner();
hmacSha1Signer.clientSharedSecret = CLIENT_SECRET;
hmacSha1Signer.tokenSharedSecret = TOKEN_SECRET;
OAuthParameters oauthParameters = new OAuthParameters();
oauthParameters.signer = hmacSha1Signer;
oauthParameters.consumerKey = CONSUMER_KEY;
oauthParameters.token = ACCESS_TOKEN;
HttpRequestFactory requestFactory = new NetHttpTransport().createRequestFactory(oauthParameters);
```

Making the fields on the new class private with setters would make the code look like this:

New code needed for HMAC-SHA256:
```
OAuthHmacSha256Signer hmacSha256Signer = new OAuthHmacSha256Signer();
hmacSha256Signer.setClientSecret(CLIENT_SECRET);
hmacSha256Signer.setTokenSecret(TOKEN_SECRET);
OAuthParameters oauthParameters = new OAuthParameters();
oauthParameters.signer = hmacSha256Signer;
oauthParameters.consumerKey = CONSUMER_KEY;
oauthParameters.token = ACCESS_TOKEN;
HttpRequestFactory requestFactory = new NetHttpTransport().createRequestFactory(oauthParameters);
```

It feels a little bit inconsistent this way, even though this is arguably the better way.  Should anything be done about `OAuthParameters` ?  I'd normally not change something if it isn't necessary.

Reply:
I made the fields private, with setter methods.